File: android/app/src/main/kotlin/io/beatscratch/beatscratch_flutter_redux/BeatScratchPlugin.kt
package io.beatscratch.beatscratch_flutter_redux

import android.content.Intent
import android.os.Handler
import io.beatscratch.beatscratch_flutter_redux.AndroidMidi.sendAllNotesOff
import io.beatscratch.beatscratch_flutter_redux.MelodyRecorder.recordingMelody
import io.beatscratch.beatscratch_flutter_redux.MelodyRecorder.recordingMelodyId
import io.beatscratch.beatscratch_flutter_redux.ScorePlayer.currentTick
import io.beatscratch.beatscratch_flutter_redux.ScorePlayer.metronomeEnabled
import io.beatscratch.beatscratch_flutter_redux.ScorePlayer.playbackMode
import io.beatscratch.beatscratch_flutter_redux.ScorePlayer.ticksPerBeat
import io.beatscratch.beatscratch_flutter_redux.hardware.MidiControllers
import io.beatscratch.beatscratch_flutter_redux.hardware.MidiDevices
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.Result
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import org.beatscratch.commands.ProtobeatsPlugin.*
import org.beatscratch.models.Music.*
import kotlin.coroutines.CoroutineContext
import io.beatscratch.beatscratch_flutter_redux.hardware.MidiDevices.name
import kotlin.concurrent.thread

object BeatScratchPlugin : MethodChannel.MethodCallHandler, CoroutineScope {
  const val internalId = "internal"
  private const val label = "BeatScratch\nAudio System"
  var handler: Handler? = null
  var methodChannel: MethodChannel? = null
    set(value) {
      field = value
      value?.setMethodCallHandler(this)
    }
  var currentScore: Score? = null
    set(value) {
      val oldValue = field
      field = value
      MidiDevices.refreshInstruments()
      tickPosition = 0
      if (value?.sectionsList?.none { it.id == currentSectionId } == true) {
        currentSectionId = value.sectionsList?.firstOrNull()?.id ?: "foo"
      }
    }
  private var currentSectionId: String? = null
  var currentSection: Section?
    get() = currentScore?.sectionsList?.firstOrNull { it.id == currentSectionId }
    set(value) {
      currentSectionId = value?.id
      PlaybackThread.unmultipliedBpm = value?.tempo?.bpm ?: 123f
    }
  var tickPosition: Int = 0
  var keyboardPart: Part? = null
  var colorboardPart: Part? = null

  private var newMelodies = mutableListOf<Melody>()

  override fun onMethodCall(call: MethodCall, result: Result) {
    when (call.method) {
      "sendMIDI"                              -> {
        AndroidMidi.sendImmediately(call.arguments as ByteArray, record = true)
        result.success(null)
      }
      "createScore", "updateSections"         -> {
        try {
          val score: Score = Score.parseFrom(call.arguments as ByteArray)
          currentSection = score.sectionsList.firstOrNull {
            it.id == currentSection?.id
          }
            ?: score.sectionsList[0]
          if (call.method == "createScore") {
            currentScore = score
            notifyMidiDevices()
          } else if (call.method == "updateSections") {
            currentScore = currentScore?.let {
              Score.newBuilder(it)
                .clearSections()
                .addAllSections(score.sectionsList)
                .build()
            }
          }
          result.success(null)
        } catch (e: Throwable) {
          logE("Failed to deserialize score", e)
          result.error("deserialize", "Failed to deserialize score", e)
        }
      }
      "createPart", "updatePartConfiguration" -> {
        try {
          val part: Part = Part.parseFrom(call.arguments as ByteArray)
          part.instrument.sendSelectInstrument()
          AndroidMidi.flushSendStream()
          if (call.method == "updatePartConfiguration") {
            currentScore?.partsList?.firstOrNull { it.id == part.id }?.let { oldPart ->
              val updatedPart = Part.newBuilder(part)
                .addAllMelodies(oldPart.melodiesList)
                .build()
              updatePart(updatedPart)
              if (keyboardPart?.id == part.id) {
                keyboardPart = part
              }
              result.success(null)
            } ?: result.error("500", "Part does not exist", "nope")
          } else {
            currentScore = currentScore?.let {
              Score.newBuilder(it)
                .addParts(part)
                .build()
            }
            result.success(null)
          }
        } catch (e: Throwable) {
          logE("Failed to deserialize part", e)
          result.error("deserialize", "Failed to deserialize part", e)
        }
      }
      "deletePart"                            -> {
        try {
          val partId = call.arguments as String
          currentScore?.let { score: Score ->
            score.partsList.firstOrNull { it.id == partId }?.let { part ->
              val partIndex = score.partsList.indexOfFirst { it.id == part.id }
              currentScore = currentScore?.let {
                Score.newBuilder(it)
                  .removeParts(partIndex)
                  .build()
              }
            }
          }
          result.success(null)
        } catch (e: Exception) {
          result.error("Cannot serialize data", null, e)
        }
      }
      "newMelody"                             -> {
        val melody: Melody = Melody.parseFrom(call.arguments as ByteArray)
        newMelodies.add(melody)
        result.success(null)
      }
      "registerMelody"                        -> {
        val registerMelody = RegisterMelody.parseFrom(call.arguments as ByteArray)
        newMelodies.firstOrNull { it.id == registerMelody.melodyId }?.let { melody ->
          currentScore?.partsList?.firstOrNull { it.id == registerMelody.partId }?.let { part ->
            val updatedPart = Part.newBuilder(part)
              .addMelodies(melody)
              .build()
            updatePart(updatedPart)
            result.success(null)
          } ?: result.error("500", "Part does not exist", "nope")
        } ?: result.error("500", "Melody must be added first", "nope")
      }
      "updateMelody"                          -> {
        val melody: Melody = Melody.parseFrom(call.arguments as ByteArray)
        if (updateMelody(melody)) {
          result.success(null)
        } else {
          result.error("500", "Part does not exist", "nope")
        }
      }
      "deleteMelody"                          -> {
        val melodyId = call.arguments as String
        currentScore?.partFor(melodyId)?.let { part ->
          val melodyIndex = part.melodiesList.indexOfFirst { it.id == melodyId }
          val updatedPart = Part.newBuilder(part)
            .removeMelodies(melodyIndex)
            .build()
          updatePart(updatedPart)
        } ?: result.error("500", "Part does not exist", "nope")
      }
      "setKeyboardPart"                       -> {
        val partId = call.arguments as String
        currentScore?.partsList?.first { it.id == partId }?.let { part ->
          keyboardPart = part
        }
        result.success(null)
      }
      "checkBeatScratchAudioStatus"                -> {
        notifyMidiDevices()
        result.success(AndroidMidi.isMidiReady)
      }
      "resetAudioSystem"                      -> {
        launch {
          AndroidMidi.resetFluidSynth()
        }
        result.success(null)
      }
      "play"                                  -> {
        val intent = Intent(MainApplication.instance, PlaybackService::class.java)
        intent.action = PlaybackService.Companion.Action.PLAY_ACTION
        MainApplication.instance.startService(intent)
        result.success(null)
      }
      "pause"                                 -> {
        val intent = Intent(MainApplication.instance, PlaybackService::class.java)
        intent.action = PlaybackService.Companion.Action.PAUSE_ACTION
        MainApplication.instance.startService(intent)
        result.success(null)
      }
      "stop"                                  -> {
        val intent = Intent(MainApplication.instance, PlaybackService::class.java)
        intent.action = PlaybackService.Companion.Action.PAUSE_ACTION
        MainApplication.instance.startService(intent)
        result.success(null)
        currentTick = 0
      }
      "setBeat"                               -> {
        val beat = call.arguments as Int
        sendAllNotesOff(immediately = true)
        currentTick = beat * ticksPerBeat
        result.success(null)
      }
      "setCurrentSection"                     -> {
        val sectionId = call.arguments as String
        currentScore?.sectionsList?.firstOrNull { it.id == sectionId }?.let { section ->
//          sendAllNotesOff(immediately = true)
          currentSection = section
          result.success(null)
          ScorePlayer.clearNonSectionActiveAttacks()
          launch {
            PlaybackService.instance?.showNotification()
          }
        } ?: result.error("500", "Section not found", "nope")
      }
      "countIn"                               -> {
        val beat = call.arguments as Int
        PlaybackThread.sendBeat()
        result.success(null)
      }
      "tickBeat"                              -> {
        PlaybackThread.sendBeat()
        result.success(null)
      }
      "setPlaybackMode"                       -> {
        val playback: Playback = Playback.parseFrom(call.arguments as ByteArray)
        playbackMode = playback.mode
        result.success(null)
      }
      "setRecordingMelody"                    -> {
        val melodyId = call.arguments as String?
        recordingMelodyId = melodyId
        result.success(null)
      }
      "setMetronomeEnabled"                   -> {
        val enabled = call.arguments as Boolean
        metronomeEnabled = enabled
        result.success(null)
      }
      "setBpmMultiplier" -> {
        val bpmMultiplier = call.arguments as Double
        PlaybackThread.bpmMultiplier = bpmMultiplier.toFloat()
        result.success(null)
      }
      "updateSynthesizerConfig" -> {
        logI("OG: $enabledSynthesizersByName")
        val config = MidiSynthesizer.parseFrom(call.arguments as ByteArray)
        val name = if (config.name == label) internalId else config.name
        enabledSynthesizersByName = if (config.enabled) {
          enabledSynthesizersByName + name
        } else {
          sendAllNotesOff(true)
          enabledSynthesizersByName - name
        }
        MidiDevices.refreshInstruments()
        logI("updated: $enabledSynthesizersByName")
      }
      else                                    -> result.notImplemented()
    }
  }

  fun updatePart(updatedPart: Part): Boolean {
    currentScore?.let { score: Score ->
      score.partsList.firstOrNull { it.id == updatedPart.id }?.let { part ->
        val partIndex = score.partsList.indexOfFirst { it.id == part.id }
        currentScore = currentScore?.let {
          Score.newBuilder(it)
            .removeParts(partIndex)
            .addParts(updatedPart)
            .build()
        }
        return true
      }
    }
    return false
  }

  fun updateMelody(melody: Melody): Boolean {
    currentScore?.let { score: Score ->
      score.partsList.firstOrNull { it.melodiesList.any { it.id == melody.id } }?.let { part ->
        val melodyIndex = part.melodiesList.indexOfFirst { it.id == melody.id }
        val partIndex = score.partsList.indexOfFirst { it.id == part.id }
        currentScore = currentScore?.let {
          val updatedPart = Part.newBuilder(part)
            .removeMelodies(melodyIndex)
            .addMelodies(melody)
          Score.newBuilder(it)
            .removeParts(partIndex)
            .addParts(updatedPart)
            .build()
        }
        return true
      }
    }
    return false
  }

  fun sendPressedMidiNotes() {
    print("kotlin: sendPressedMidiNotes")
    val midiNotes: MidiNotes = MidiNotes.newBuilder()
      .addAllMidiNotes(MidiControllers.pressedNotes).build()
    handler?.post {
      methodChannel?.invokeMethod("sendPressedMidiNotes", midiNotes.toByteArray())
    }
  }

  fun notifyRecordedMelody() {
    recordingMelody?.let { recordingMelody ->
      handler?.post {
        methodChannel?.invokeMethod("notifyRecordedMelody", recordingMelody.toByteArray())
      }
    }
  }

  fun notifyBeatScratchAudioAvailable() {
    handler?.post {
      methodChannel?.invokeMethod("notifyBeatScratchAudioAvailable", AndroidMidi.isMidiReady)
    }
  }

  fun notifyPlayingBeat() {
    handler?.post {
      val beat: Int = currentTick / ticksPerBeat
      methodChannel?.invokeMethod("notifyPlayingBeat", beat)
    }
  }

  fun notifyPaused() {
    handler?.post {
      methodChannel?.invokeMethod("notifyPaused", null)
    }
  }

  fun notifyCountInInitiated() {
    handler?.post {
      methodChannel?.invokeMethod("notifyCountInInitiated", null)
    }
  }

  fun notifyCurrentSection() {
    handler?.post {
      methodChannel?.invokeMethod("notifyCurrentSection", currentSection!!.id)
    }
  }

  fun notifyStartedSection() {
    handler?.post {
      methodChannel?.invokeMethod("notifyStartedSection", currentSection!!.id)
    }
  }

  fun notifyBpmMultiplier() {
    handler?.post {
      methodChannel?.invokeMethod("notifyBpmMultiplier", PlaybackThread.bpmMultiplier)
    }
  }

  fun notifyUnmultipliedBpm() {
    handler?.post {
      methodChannel?.invokeMethod("notifyUnmultipliedBpm", PlaybackThread.unmultipliedBpm)
    }
  }

  var enabledSynthesizersByName by stringSetPref("enabledSynthesizersByName", setOf(internalId))
    private set
  fun notifyMidiDevices() {
    handler?.post {
      val midiDevices = org.beatscratch.commands.ProtobeatsPlugin.MidiDevices.newBuilder()
        .addAllControllers(MidiDevices.controllers.map {
          MidiController.newBuilder()
            .setId(it.info.id.toString())
            .setName(it.info.name)
            .setEnabled(true)
            .build()
        })
        .addSynthesizers(MidiSynthesizer.newBuilder()
          .setId(internalId)
          .setName(label)
          .setEnabled(enabledSynthesizersByName.contains(internalId))
          .build())
        .addAllSynthesizers(MidiDevices.synthesizers.map {
          MidiSynthesizer.newBuilder()
            .setId(it.info.id.toString())
            .setName(it.info.name)
            .setEnabled(enabledSynthesizersByName.contains(it.info.name))
            .build()
        }).build()
      methodChannel?.invokeMethod("notifyMidiDevices", midiDevices.toByteArray())
    }
  }

  fun notifyScoreUrlOpened(url: String) {
    handler?.post {
      try {
        methodChannel!!.invokeMethod("notifyScoreUrlOpened", url, object : Result {
          override fun success(result: Any?) {}

          override fun error(errorCode: String?, errorMessage: String?, errorDetails: Any?) {
            thread {
              Thread.sleep(500)
              notifyScoreUrlOpened(url)
            }
          }

          override fun notImplemented() {
            thread {
              Thread.sleep(500)
              notifyScoreUrlOpened(url)
            }
          }
        })
      } catch(t: Throwable) {
        thread {
          Thread.sleep(500)
          notifyScoreUrlOpened(url)
        }
      }
    }
  }

  override val coroutineContext: CoroutineContext = Dispatchers.Default
}

File: macos/Runner/BeatScratch/BeatScratchPlugin.swift
//
//  BeatScratchPlugin.swift
//  Runner
//
//  Created by Jon Latane on 4/18/20.
//  Copyright © 2020 The Flutter Authors. All rights reserved.
//

import Foundation
#if canImport(FlutterMacOS)
import FlutterMacOS
#elseif canImport(Flutter)
import Flutter
#endif

class BeatScratchPlugin {
  // Globally accessible
  static let sharedInstance = BeatScratchPlugin()
  var score: Score = Score()
  var channel: FlutterMethodChannel?
  
  private init() {}
  
  private var newMelodies: Array<Melody> = []

  func attach(channel: FlutterMethodChannel) {
    let conductor = Conductor.sharedInstance
    self.channel = channel
    channel.setMethodCallHandler { (call, result) in
      print("Call from Swift: " + call.method)
      do {
        switch call.method {
        case "sendMIDI":
          let data = (call.arguments as! FlutterStandardTypedData).data
          let args = [UInt8](data)
          let parsedBytes = conductor.parseMidi(args, record: true)
          if parsedBytes > 0 {
            result(nil)
          } else {
            result(FlutterMethodNotImplemented)
          }
          break
        case "createScore", "updateSections":
          let score = try Score(serializedData: (call.arguments as! FlutterStandardTypedData).data)
          BeatScratchScorePlayer.sharedInstance.currentSection = score.sections.first {
            $0.id == BeatScratchScorePlayer.sharedInstance.currentSection.id
            } ?? score.sections.first!
          if call.method == "createScore" {
            self.score = score
          } else if call.method == "updateSections" {
            self.score.sections = score.sections
          }
          result(nil)
          break
        case "createPart", "updatePartConfiguration":
          var part = try Part(serializedData: (call.arguments as! FlutterStandardTypedData).data)
          conductor.setMIDIInstrument(channel: part.instrument.midiChannel, midiInstrument: part.instrument.midiInstrument)
          conductor.setVolume(channel: part.instrument.midiChannel, volume: Double(part.instrument.volume))
          if call.method == "updatePartConfiguration" {
            if let existingPart = self.score.parts.first(where: {$0.id == part.id}) {
              do {
                try part.melodies = existingPart.melodies
                try self.score.parts.removeAll(where: {$0.id == part.id})
                try self.score.parts.append(part)
              } catch {
                print("updatePartConfiguration error: \(error)")
              }
              result(nil)
            } else {
              result(FlutterError(code: "500", message: "Part does not exist", details: "nope"))
            }
          } else {
            self.score.parts.append(part)
            result(nil)
          }
          break
        case "deletePart":
          let partId = call.arguments as! String
          self.score.parts.removeAll { $0.id == partId }
          result(nil)
          break
        case "newMelody":
          let melody = try Melody(serializedData: (call.arguments as! FlutterStandardTypedData).data)
          self.newMelodies.append(melody)
          result(nil)
          break
        case "registerMelody":
          let registerMelody = try RegisterMelody(serializedData: (call.arguments as! FlutterStandardTypedData).data)
          if let melody: Melody = self.newMelodies.first(where: { $0.id == registerMelody.melodyID }) {
            if var part: Part = self.score.parts.first(where: { $0.id == registerMelody.partID }) {
              part.melodies.append(melody)
              self.score.parts.removeAll { $0.id == part.id }
              self.score.parts.append(part)
              self.newMelodies.removeAll { $0.id == registerMelody.melodyID }
              result(nil)
            } else {
              result(FlutterError(code: "500", message: "Part must be added first", details: "nope"))
            }
          } else {
            result(FlutterError(code: "500", message: "Melody must be added first", details: "nope"))
          }
          break
        case "updateMelody":
          let melody = try Melody(serializedData: (call.arguments as! FlutterStandardTypedData).data)
          if var part: Part = self.score.parts.first(where:{
            $0.melodies.contains(where: { $0.id == melody.id })
          }) {
            part.melodies.removeAll { $0.id == melody.id }
            part.melodies.append(melody)
          } else {
            result(FlutterError(code: "500", message: "Melody not found in any Part", details: "nope"))
          }
          self.newMelodies.append(melody)
          result(nil)
          break
        case "deleteMelody":
          let melodyId = call.arguments as! String
          self.score.parts.forEach {
            var part = $0
            part.melodies.removeAll { $0.id == melodyId }
            self.score.parts.removeAll { $0.id == part.id }
            self.score.parts.append(part)
          }
          result(nil)
          break
        case "setKeyboardPart":
          let partId = call.arguments as! String
          if let part = self.score.parts.first(where: {$0.id == partId}) {
            let channel = Int(part.instrument.midiChannel)
            BeatScratchMidiListener.sharedInstance.conductorChannel = channel
          }
          break
        case "checkSynthesizerStatus":
          result(Conductor.sharedInstance.samplersInitialized)
          break
        case "resetAudioSystem":
          result(nil)
          break
        case "play":
          BeatScratchPlaybackThread.sharedInstance.playing = true
          result(nil)
          break
        case "pause":
          BeatScratchPlaybackThread.sharedInstance.stopped = true
          result(nil)
          break
        case "stop":
          BeatScratchPlaybackThread.sharedInstance.stopped = true
          BeatScratchScorePlayer.sharedInstance.currentTick = 0
          result(nil)
          break
        case "setBeat":
          let beat = call.arguments as! Int
          Conductor.sharedInstance.stopPlayingNotes()
          BeatScratchScorePlayer.sharedInstance.currentTick = Int64(beat) * Int64( BeatScratchPlaybackThread.ticksPerBeat)
          result(nil)
          break
        case "setCurrentSection":
          let sectionId = call.arguments as! String
          if let section = self.score.sections.first(where: { $0.id == sectionId }) {
            BeatScratchScorePlayer.sharedInstance.currentSection = section
            Conductor.sharedInstance.stopPlayingNotes()
            result(nil)
          } else {
            result(FlutterError(code: "500", message: "Section not found", details: "nope"))
          }
          break
        case "countIn":
          let countInBeat = call.arguments as! Int
          BeatScratchPlaybackThread.sharedInstance.sendBeat()
          result(nil)
          break
        case "tickBeat":
          BeatScratchPlaybackThread.sharedInstance.sendBeat()
          result(nil)
          break
        case "setPlaybackMode":
          let playback: Playback = try Playback(serializedData: (call.arguments as! FlutterStandardTypedData).data)
          BeatScratchScorePlayer.sharedInstance.playbackMode = playback.mode
          result(nil)
          break
        case "setRecordingMelody":
          let melodyId: String? = call.arguments as! String?
          MelodyRecorder.sharedInstance.recordingMelodyId = melodyId
          result(nil)
          break
        case "setMetronomeEnabled":
          let metronomeEnabled: Bool = call.arguments as! Bool
          BeatScratchScorePlayer.sharedInstance.metronomeEnabled = metronomeEnabled
          result(nil)
          break
        default:
          result(FlutterMethodNotImplemented)
          break
        }
      } catch {
        print("Failed to process Platform Channel call in macOS. Error: " + error.localizedDescription)
        result(FlutterError())
      }
    }
  }
  
  func sendPressedMidiNotes() {
    do {
//      print("swift: sendPressedMidiNotes")
      var midiNotes = MidiNotes()
      midiNotes.midiNotes = BeatScratchMidiListener.sharedInstance.pressedNotes.map { UInt32($0) }
      channel?.invokeMethod("sendPressedMidiNotes", arguments: try midiNotes.serializedData())
    } catch {
      print("Failed to sendPressedMidiNotes: \(error)")
    }
  }
  
  func sendRecordedMelody() {
    do {
      if let recordingMelody = MelodyRecorder.sharedInstance.recordingMelody {
        if recordingMelody.midiData.data.count > 0 {
          channel?.invokeMethod("sendRecordedMelody", arguments: try recordingMelody.serializedData())
        }
      }
    } catch {
      print("Failed to sendRecordedMelody: \(error)")
    }
  }
  
  func setSynthesizerAvailable() {
    channel?.invokeMethod("setSynthesizerAvailable", arguments: Conductor.sharedInstance.samplersInitialized)
  }
  
  func notifyPlayingBeat() {
    let beat: Int = Int(BeatScratchScorePlayer.sharedInstance.currentTick / Int64(BeatScratchPlaybackThread.ticksPerBeat))
    channel?.invokeMethod("notifyPlayingBeat", arguments: beat)
  }
  
  func notifyPaused() {
    channel?.invokeMethod("notifyPaused", arguments: nil)
  }

  func notifyCountInInitiated() {
    channel?.invokeMethod("notifyCountInInitiated", arguments: nil)
  }
  
  func notifyCurrentSection() {
    let section = BeatScratchScorePlayer.sharedInstance.currentSection
    channel?.invokeMethod("notifyCurrentSection", arguments: section.id)
  }
}
